presto-main/src/main/java/com/facebook/presto/operator/DecodedBlockNode.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.Block;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+class DecodedBlockNode
```

Comment:
Any document about "DecodedBlockNode" that can help reader understand the design of Aria Partition? :) 

Again, I am not sure if this class should be put into package `com.facebook.presto.operator`

Reply:
> Any document about "DecodedBlockNode" that can help reader understand the design of Aria Partition? :)
Added some explanation and example in the code

> Again, I am not sure if this class should be put into package `com.facebook.presto.operator`
Where do you think it belongs to?

Reply:
> Any document about "DecodedBlockNode" that can help reader understand the design of Aria Partition? :)

Hi @wenleix How about these code comments?

```
/**
 * A tree structure to represent a decoded block. Dictionary/Rle blocks will be kept as a
 * separate node because we need the position mappings. For example, a block of map<int, bigint> where the key
 * is dictionary encoded will result in the following tree structure:
 * 
 *             ColumnarMap
 *               /      \
 *    DictionaryBlock   LongArrayBlock
 *          |
 *     IntArrayBlock
 */
class DecodedBlock
```

 ```
   /**
     * Flatten the block and convert the nested-typed block into ColumnarArray/Map/Row.
     * For performance considerations we decode the block only once for each block instead of each batch.
     * @return A tree structure that contains the decoded block
     */
    @VisibleForTesting
    static DecodedBlock decodeBlock(BlockFlattener flattener, Block block)
```

Reply:
@wenleix 

> Again, I am not sure if this class should be put into package `com.facebook.presto.operator`

I think it makes sense for them to belong to the same package as the operator and BlockEncodingBuffers. First of all BlockEncodingBuffers is better not to be put in presto-spi as we don't want to pollute spi, and it's only used by OptmizedPartitionedOutputOperator. Then the DecodeBlockNode is also a hack only used by OptmizedPartitionedOutputOperator. If later on other operators can make use of them we can move these classes to presto-spi then.

Reply:
@yingsu00 : Yup. These comments would be great to add.